R/weight.R

```diff
@@ -41,33 +43,216 @@ weight <- function (x) {
+#' #To append a weight to existing weights
+#' weightVariables(ds) <- c(weightVariables(ds), "weight3")
+#' #To remove a weight
+#' weightVariables(ds) <- setdiff(weightVariables(ds), "weight3")
+#' }
+#' @name weightVaraibles
+#' @aliases is.weightVariable<- weightVariables<-
+#' 
+#' @export
+NULL
+
+#' @rdname weightVariables
+#' @export
+setMethod("weightVariables", "CrunchDataset",
+    function(x) weightVariables(allVariables(x)))
+
+#' @rdname weightVariables
+#' @export
+setMethod("weightVariables", "VariableCatalog", function(x) {
     ## Get weight variable order
     ord <- VariableOrder(crGET(shojiURL(x, "orders", "weights")))
     ## Subset weight variable catalog with it
     vars <- x[ord]
     ## Return the names/aliases
     if (length(vars)) {
         return(sort(vapply(index(vars), vget(namekey(x)), character(1),
-            USE.NAMES=FALSE)))
+            USE.NAMES = FALSE)))
     } else {
         return(c())
     }
 })
 
+#' @rdname weightVariables
+#' @export
+setMethod("weightVariables<-", "CrunchDataset", function(x, value) {
+    weightVariables(allVariables(x)) <- value
+    x <- refresh(x)
+    return(x)
+})
+
+#' @rdname weightVariables
+#' @export
+setMethod("weightVariables<-", "VariableCatalog", function(x, value) {
+    modifyWeightVariables(x, value, type = "append")
+})
+
+#' @rdname modifyWeightVariables
+#' @export
+setMethod("modifyWeightVariables", "CrunchDataset", function(x, vars, type = "append") {
+    varcat <- allVariables(x)
+    modifyWeightVariables(varcat, vars, type)
+})
+
+#' @rdname modifyWeightVariables
+#' @export
+setMethod("modifyWeightVariables", "VariableCatalog", function(x, vars, type = "append") {
+    modifyWeightVariables(x, vars, type)
+})
+
+#' modifyWeightVariables
+#'
+#' Change which variables can be set as a dataset's weight.
+#'
+#' @param x a dataset or variable catalog
+#' @param vars Variables to add or remove  this can be a numeric Crunch variable,
+#' list of numeric Crunch variables or a character vector with the aliases of numeric Crunch variables.
+#' @param type a character string determining how the weightVariables
+#' will be modified:
+#' - `"append"` : add `vars` to the current weight variables
+#' - `"remove"` : remove `vars` from the current list of weight variables
+#' - `"replace"`: replace the current weight variables with `vars`
+#'
+#' @export
+modifyWeightVariables <- function(x, vars, type = "append") {
+        current <- crGET(shojiURL(x, "orders", "weights"))
+    new <- current
+    if (is.null(vars)) {
+        new$graph <- NULL
+        type = "remove"
+    } else {
+        if (is.variable(vars) || (length(vars) == 1) & !is.character(vars)) {
+            vars <- list(vars)
+        }
+        if (is.character(vars)) {
+            ds <- loadDataset(datasetReference(self(x)))
+            vars <- lapply(vars, function(v){
+                if (v %in% names(x)) {
+                    ds[[v]]
+                } else {
+                    v
```

Comment:
Why this "else" case? Under what circumstance would v not be in names(x) and that's ok?

Reply:
My thought was if someone assigned something like `is.weight(ds) <- c("weight1", "weight23")` where weight23 wasn't a variable I would want to error by saying "weight 23 is not a valid variable" so I need to preserve the name that was passed to trigger the correct errors later on.